### ISSUE TITLE
studip-fuse fails to start: No module named 'studip_fuse.launcher.aioimpl'

### ISSUE THREAD

**Issue by n-st** [N-Coder/studip-fuse#14]
`studip-fuse` fails to start after installing it in a virtualenv with the `pip` command from the README:

```
% python3 -m venv /home/nst/bin/py/studip-fuse
% source /home/nst/bin/py/studip-fuse/bin/activate
% pip3 install git+https://github.com/N-Coder/studip-api git+https://github.com/N-Coder/studip-fuse
…
% studip-fuse mueller123 /home/nst/StudIP.fuse --pwfile=- --foreground
Traceback (most recent call last):
  File "/home/nst/bin/py/studip-fuse/bin/studip-fuse", line 11, in <module>
    load_entry_point('studip-fuse==3.0', 'console_scripts', 'studip-fuse')()
  File "/home/nst/bin/py/studip-fuse/lib/python3.7/site-packages/pkg_resources/__init__.py", line 487, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/home/nst/bin/py/studip-fuse/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2728, in load_entry_point
    return ep.load()
  File "/home/nst/bin/py/studip-fuse/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2346, in load
    return self.resolve()
  File "/home/nst/bin/py/studip-fuse/lib/python3.7/site-packages/pkg_resources/__init__.py", line 2352, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/nst/bin/py/studip-fuse/lib/python3.7/site-packages/studip_fuse/launcher/main.py", line 5, in <module>
    import studip_fuse.launcher.aioimpl.asyncio as aioimpl_asyncio
ModuleNotFoundError: No module named 'studip_fuse.launcher.aioimpl'
```

This is using Python 3.7.2 on Arch Linux.

**Comment by N-Coder** [comment]
Unfortunately, I released version 3.0 once all the actual code was there, but didn't check the documentation and install scripts.

The package structure was changed in transition to version 3.0, without correctly taking care of the packaging process, so `pip` (and setuptools) are currently unable to install version 3.0 correctly in production mode. If you really want the stable version, use the `--editable` flag with `pip` or alternatively `develop` instead of `install` with an already checked-out `setup.py`.

The ['dev' branch](https://github.com/N-Coder/studip-fuse/tree/dev) also already has an updated [README.md](https://github.com/N-Coder/studip-fuse/blob/dev/README.md) (plus some [minor fixes](https://github.com/N-Coder/studip-fuse/compare/dev), including a fix to this issue in https://github.com/N-Coder/studip-fuse/commit/1c6995bed2774b5ea27781e9b76cf2b521576c20), that doesn't list the dependency on the outdated studip-api package (which is no longer required thanks to the official [Stud.IP REST API](https://hilfe.studip.de/develop/Entwickler/RESTAPI). The [studip-api repo](https://github.com/N-Coder/studip-api/) is now deprecated and archived.).
I still want to fix some small ToDos before releasing 3.1, but there should be no breaking changes on `dev` until then.

*tl;dr* Install from the dev branch until version 3.1 is released by using:
`pip3 install git+https://github.com/N-Coder/studip-fuse@dev`

**Comment by leon1995** [comment]
I’ve just got the same error, when following the README how to install studip-fuse. Then I followed your instruction and installed the dev branch. Then I got
```
$ studip-fuse -h
Traceback (most recent call last):
  File "./studip-fuse", line 7, in <module>
    from studip_fuse.launcher.main import main
  File "/home/username/.local/lib/python3.6/site-packages/studip_fuse/launcher/main.py", line 5, in <module>
    import studip_fuse.launcher.aioimpl.asyncio as aioimpl_asyncio
  File "/home/username/.local/lib/python3.6/site-packages/studip_fuse/launcher/aioimpl/asyncio/__init__.py", line 1, in <module>
    from studip_fuse.launcher.aioimpl.asyncio.aiohttp_client import AiohttpClient as HTTPClient, AiohttpDownload as Download, AuthenticatedClientRequest
  File "/home/username/.local/lib/python3.6/site-packages/studip_fuse/launcher/aioimpl/asyncio/aiohttp_client.py", line 9, in <module>
    from contextlib import AsyncExitStack
ImportError: cannot import name 'AsyncExitStack'
```
I’m using Ubuntu 18.04 and python 3.6

Hope you can help me :)

**Comment by n-st** [comment]
Looks like that's a new feature of Python 3.7: https://docs.python.org/3/whatsnew/3.7.html#contextlib
Afaik @N-Coder uses Fedora (which ships Python 3.7), so no-one noticed the incompatibility…

@leon1995 Out of curiousity, how come a TUHH student is using a Passau-specific StudIP wrapper? :smile:

**Comment by N-Coder** [comment]
Try reinstalling the dev Version at the latest commit 9a9e496a220314502bdd4c41c6e5b573b6238dc3, which should fix the compatibility issue. Unfortunately, there is  [no single backport in contextlib2 for async functions yet](https://github.com/jazzband/contextlib2/issues/12), so I'm now using [python-trio/async_generator](https://github.com/python-trio/async_generator) and [sorcio/async_exit_stack](https://github.com/sorcio/async_exit_stack).

Automated testing with different Python Versions and OSes is what keeps me from finally releasing version 3.1 of Stud.IP-Fuse, so keep us posted on any further problems. :wink: 

> @leon1995 Out of curiousity, how come a TUHH student is using a Passau-specific StudIP wrapper?

Well, it's not that specific to Passau. Except for the scripted Shibboleth login, everything should be compatible with any sufficiently recent Stud.IP version. Speaking of login: could you PM me some details on how the login procedure (just the procedure - no credentials :stuck_out_tongue:) works for your university and whether anything breaks? I'd be glad to help with ensuring compatibility with further Stud.IP instances.

**Comment by leon1995** [comment]
Thx for the quick answer. I reinstalled using
`$ pip3 install git+https://github.com/N-Coder/studip-fuse@dev --user --upgrade`
But now I get the old error again:
```
$ studip-fuse -h
Traceback (most recent call last):
  File "./studip-fuse", line 7, in <module>
    from studip_fuse.launcher.main import main
  File "/home/leon/.local/lib/python3.6/site-packages/studip_fuse/launcher/main.py", line 5, in <module>
    import studip_fuse.launcher.aioimpl.asyncio as aioimpl_asyncio
  File "/home/leon/.local/lib/python3.6/site-packages/studip_fuse/launcher/aioimpl/asyncio/__init__.py", line 2, in <module>
    from studip_fuse.launcher.aioimpl.asyncio.main_loop import setup_asyncio_loop as setup_loop
  File "/home/leon/.local/lib/python3.6/site-packages/studip_fuse/launcher/aioimpl/asyncio/main_loop.py", line 13, in <module>
    import studip_fuse.launcher.aioimpl.asyncio as aioimpl_asyncio
AttributeError: module 'studip_fuse.launcher.aioimpl' has no attribute 'asyncio'
```

@n-st I’m trying to find different ways to interact with stud-ip, because the mobile website is very **very** bad. I know there is an api for stud-ip but it has to be enabled by the system administrator and I don’t if my very **very** good _technical_ university is willing to enable it. There is an official app at the App Store but it only supports few universities.

@N-Coder Ofc I want to help. Can you give me a hint what do you mean with login procedure? I’m not sure what do you mean. EDIT: I didn’t even know that there are different login procedures for different universities using stud-ip

**Comment by N-Coder** [comment]
Your [login](https://e-learning.tuhh.de/studip/index.php?again=yes) seems to be the built-in one, so it should work with `--login-method basic`. Using this method (i.e. HTTP basic auth), you should also be able to access the regular API. Instead of HTTP basic auth, an existing session where you are logged in should also be sufficient, so could just try accessing [this page](https://e-learning.tuhh.de/studip/api.php/discover) in your browser and send me the response via mail? (using the address from [here](https://github.com/N-Coder/studip-fuse/blob/master/studip_fuse/__init__.py))

I'm currently setting up a VM to investigate your problem.

**Comment by N-Coder** [comment]
Would you try again using 1239f3e3d4d4268e647685c36ab02031b266bd02?

Note to self: Ubuntu 16.04 is no longer supported, because it [only ships python 3.5.2](https://packages.ubuntu.com/xenial-updates/python3.5), but [aiohttp requires >= 3.5.3](https://github.com/aio-libs/aiohttp/blob/master/docs/faq.rst#why-is-python-3-5-3-the-lowest-supported-version). Ubuntu 18.04 and Python 3.6.7 have been successfully tested.

**Comment by leon1995** [comment]
It works. But when im trying to
`$ studip-fuse MYUSERNAME ./studip/ --pwfile=- --foreground --login-method=basic`
and i enter the password the authentication failes

EDIT: I guess because it tries on studip.uni-passau.de

EDIT2: using `--studip-url="https://e-learning.tuhh.de/studip/api.php/"` returns 403, Route not activated
